mlir/lib/Analysis/DataFlow/SparseAnalysis.cpp

```diff
@@ -609,12 +609,27 @@ void AbstractSparseBackwardDataFlowAnalysis::visitRegionSuccessors(
            *getLatticeElementFor(getProgramPointAfter(op), input));
       unaccounted.reset(operand.getOperandNumber());
     }
+
+    if (successor.isParent())
+      continue;
+    MutableArrayRef<BlockArgument> arguments =
+        successor.getSuccessor()->getArguments();
+    for (BlockArgument argument : arguments) {
+      // Visit property blockArgument of RegionBranchOp which isn't "control
+      // flow block arguments". For example, the IV of a loop.
+      if (!llvm::is_contained(inputs, argument)) {
+        regionArguments.push_back(argument);
+      }
+    }
   }
   // All operands not forwarded to regions are typically parameters of the
   // branch operation itself (for example the boolean for if/else).
   for (int index : unaccounted.set_bits()) {
     visitBranchOperand(op->getOpOperand(index));
   }
+  for (BlockArgument argument : regionArguments) {
+    visitNonControlFlowArguments(argument);
+  }
```

Comment:
It looks like `regionArguments` is a flat list containing arguments from _all_ successor regions indistinguishably. While it is possible to recover the owning region from `BlockArgument`, I'd consider aligning this method on `visitNonControlFlowArguments` method of the forward analysis that takes a reference to `RegionSuccessor` and a _list_ of arguments -- as the name indicates -- for that specific successor.